src/nbl/ui/CWindowWin32.cpp

```diff
@@ -1,184 +1,62 @@
-				inputDevices[1].usUsagePage = HID_USAGE_PAGE_GENERIC;
-				inputDevices[1].usUsage = HID_USAGE_GENERIC_MOUSE;
-
-				inputDevices[2].hwndTarget = m_params.nativeWindow;
-				inputDevices[2].dwFlags = RIDEV_DEVNOTIFY | RIDEV_INPUTSINK;
-				inputDevices[2].usUsagePage = HID_USAGE_PAGE_GENERIC;
-				inputDevices[2].usUsage = HID_USAGE_GENERIC_KEYBOARD;
-
-				inputDevices[3].hwndTarget = m_params.nativeWindow;
-				inputDevices[3].dwFlags = RIDEV_DEVNOTIFY | RIDEV_INPUTSINK;
-				inputDevices[3].usUsagePage = HID_USAGE_PAGE_GAME;
-				inputDevices[3].usUsage = HID_USAGE_GENERIC_JOYSTICK;
-
-				inputDevices[4].hwndTarget = m_params.nativeWindow;
-				inputDevices[4].dwFlags = RIDEV_DEVNOTIFY | RIDEV_INPUTSINK;
-				inputDevices[4].usUsagePage = HID_USAGE_PAGE_GAME;
-				inputDevices[4].usUsage = HID_USAGE_GENERIC_GAMEPAD;
-
-				RegisterRawInputDevices(inputDevices, INPUT_DEVICES_COUNT, sizeof(RAWINPUTDEVICE));
-			}
-			m_params.nativeWindow = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
-				realWidth, realHeight, NULL, NULL, hinstance, NULL);
-			if ((m_params.flags & CWindowWin32::ECF_HIDDEN) == 0)
-				ShowWindow(m_params.nativeWindow, SW_SHOWNORMAL);
-			UpdateWindow(m_params.nativeWindow);
-
-			// fix ugly ATI driver bugs. Thanks to ariaci
-			// TODO still needed?
-			MoveWindow(m_params.nativeWindow, windowLeft, windowTop, realWidth, realHeight, TRUE);
-		}
-		void setParams(SParams&& params) override { m_params = static_cast<SCreateWindowParams&&>(params); }
+		static inline constexpr E_REQUEST_TYPE type = ERT;
 	};
-	struct SDestroyWindowRequest : SRequest
+	struct SRequestParams_CreateWindow : SRequestParamsBase<ERT_CREATE_WINDOW>
 	{
-		SDestroyWindowRequest(CWindowWin32::native_handle_t wnd) : m_params{ wnd } {}
-		struct SDestroyWindowParams : SParams
-		{
-			CWindowWin32::native_handle_t window;
-		} m_params;
-		void process() override
+		SRequestParams_CreateWindow(int32_t _x, int32_t _y, uint32_t _w, uint32_t _h, CWindowWin32::E_CREATE_FLAGS _flags, CWindowWin32::native_handle_t wnd, const std::string_view& caption):
+		x(_x), y(_y), width(_w), height(_h), flags(_flags), nativeWindow(wnd)
 		{
-			DestroyWindow(m_params.window);
+			std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
+			windowCaption = converter.from_bytes(std::string(caption));
 		}
-		void setParams(SParams&& params) override { m_params = static_cast<SDestroyWindowParams&&>(params); }
+		int32_t x, y;
+		uint32_t width, height;
+		CWindowWin32::E_CREATE_FLAGS flags;
+		CWindowWin32::native_handle_t nativeWindow;
+		std::wstring windowCaption;
 	};
+	struct SRequestParams_DestroyWindow : SRequestParamsBase<ERT_DESTROY_WINDOW>
+	{
+		CWindowWin32::native_handle_t nativeWindow;
+	};
+	struct SRequest : system::impl::IAsyncQueueDispatcherBase::request_base_t
+	{
+		E_REQUEST_TYPE type;
+		std::variant<SRequestParams_CreateWindow, SRequestParams_DestroyWindow> params;
```

Comment:
you dont need variant if the `type` is already a member above.

Reply:
why not? its just convenient

Reply:
its not convenient because it can throw https://en.cppreference.com/w/cpp/utility/variant/bad_variant_access and and my aim is for the engine to not have RTTI nor any exceptions

Reply:
done